src/Primitives/src/StringValues.cs

```diff
@@ -218,25 +219,37 @@ static string GetJoinedStringValueFromArray(string[] values)
 
         public string[] ToArray()
         {
-            return GetArrayValue() ?? Array.Empty<string>();
+            // The same idea as GetArrayValue with slightly different semantics: This method always returns a copy of the array (just like the IEnumerable.ToArray extension method).
+
+            // Take local copy of _values so type checks remain valid even if the StringValues is overwritten in memory.
+            var value = _values;
+            switch (value)
+            {
+                case string[] values:
+                    return (string[]) values.Clone();
+                case null:
+                    return Array.Empty<string>();
+                default:
+                    // value not array, can only be string
+                    return new[] { Unsafe.As<string>(value) };
+            }
         }
 
         private string[] GetArrayValue()
         {
+            // The same idea as ToArray with slightly different semantics: This method reuses the existing array if possible.
+
             // Take local copy of _values so type checks remain valid even if the StringValues is overwritten in memory
             var value = _values;
-            if (value is string[] values)
-            {
-                return values;
-            }
-            else if (value != null)
+            switch (value)
```

Comment:
The `if` generates better asm than the `switch` [sharplab.io](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYmpBkIYBvGgzNNdzOgG0Aug1kB3XhjAALABQB5YACsYYBgMCtiSHDAAlAym5ibU5gkMuM6ubgweIWGRMYnGObnmYHgwTOQ29pnhuCD5BYnEAOzpVnZRlarCkhBcMB4RGvF1hcUMXBySkjWDQ/VNAIJQUNgAniwAouIYywA8VgB8fQMzZgAmMABm2ONYtccA9HfBoeGjEEHYiytoDEVcDN2SZYMYAlKy3GaNUYwRx2YwMACqXFw2HOgjmuF2ZQO7SiAF8jgVcfkidNoqSdKVygwAJLnbx+AJBHFkhJxOq8c7pdoMXi4Smw9q4CLgvKkuqQjwtWwRQWdbq9foikl1GCSXAlDlc54lACEAF5RuNJMKxbERQlIT0YfYjAikSi0Rj9hltXiCbllQVVeqWQU2cdLUa1AklcSaLigA)

Reply:
Maybe, but it is actually measurably better? If not, the switch is a easier to read / more expressive. I try to make a benchmark on it.

Reply:
If you are benchmarking 😉 then...
```csharp
return values.AsSpan().ToArray()
```
might be better than
```csharp
return (string[])values.Clone();
```

Reply:
I can give it a shot :D

Reply:
It looks like the switch is indeed faster in most workloads (except when the payload is null), at least on my machine, although I would say the difference is negligible:

> BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17763.529 (1809/October2018Update/Redstone5)
> Intel Core i5-6440HQ CPU 2.60GHz (Skylake), 1 CPU, 4 logical and 4 physical cores
> .NET Core SDK=3.0.100-preview5-011568
>   [Host]     : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT
>   DefaultJob : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT


| Method |     Data |      Mean |     Error |    StdDev |
|------- |--------- |----------:|----------:|----------:|
| Switch |        ? _(null)_ |  3.727 ns | 0.0643 ns | 0.0601 ns |
|     If |        ? _(null)_ |  3.692 ns | 0.0730 ns | 0.0683 ns |
| Switch | Array[2] | 73.998 ns | 1.2785 ns | 1.1333 ns |
|     If | Array[2] | 76.917 ns | 1.9549 ns | 1.8286 ns |
| Switch |      abc | 11.149 ns | 0.4064 ns | 0.5140 ns |
|     If |      abc | 12.266 ns | 0.2307 ns | 0.2158 ns |

Reply:
#### `AsSpan().ToArray()` is indeed about twice as fast as `Array.Clone()`

> BenchmarkDotNet=v0.11.5, OS=Windows 10.0.17763.529 (1809/October2018Update/Redstone5)
> Intel Core i5-6440HQ CPU 2.60GHz (Skylake), 1 CPU, 4 logical and 4 physical cores
> .NET Core SDK=3.0.100-preview5-011568
>   [Host]     : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT
>   DefaultJob : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT


|      Method |     Mean |     Error |    StdDev |
|------------ |---------:|----------:|----------:|
|       Clone | 72.96 ns | 0.6034 ns | 0.5644 ns |
| SpanToArray | 27.97 ns | 0.3408 ns | 0.3188 ns |

It is nice to know this, thanks for the idea. This pretty much makes it the default go-to array clone solution on .NET Core then. I will update the code in the pull request and write about this somewhere.

Reply:
@benaadams Array.Clone, another one for the CoreFX improvement pile ;)

Reply:
Is it really faster to go through `.AsSpan().ToArray()` vs doing something like?
```
var copy= new T[orig.Length];
Array.Copy(orig, 0, copy, 0, copy.Length);
return copy;
```

Reply:
Interestingly yes. 
Check out my blog article where I test all kinds of different methods of copying an array: https://marcelltoth.net/article/41/fastest-way-to-copy-a-net-array it was inspired by this review.
The usual use case we have here is where the array is not empty but it only contains a handful of elements. In that scenario by my tests `Array.Copy` performs about 2x as slow. (See the second result set.)